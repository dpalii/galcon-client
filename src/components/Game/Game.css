.game {
    width: 100vw;
    height: 100vh;
    position: fixed;
    display: flex;
    flex-direction: column;
}

.game-map {
    flex: 1;
    background-color: #000;
    width: 100%;
    height: 100%;
}

.header {
    display: flex;
    flex-direction: row;
    padding: 5px;
    justify-content: flex-end;
    background-color: var(--bg-secondary);
}

.header .btn {
    margin-left: 10px;
}

.game-id {
    padding: 10px 0;
}
