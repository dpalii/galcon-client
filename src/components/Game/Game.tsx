import { useContext, useEffect, useState } from 'react'
import './Game.css'
import { SocketContext } from '../../contexts/SocketContext'
import { GameState, IncomingEvents, MapData, OutgoingEvents, User } from '../../types'
import { GameMap } from '../GameMap/GameMap'
import { LobbyModal } from '../LobbyModal/LobbyModal'
import { SummaryModal } from '../SummaryModal/SummaryModal'

export interface GameProps {
    gameId: string
    players: User[]
    gameFinished: () => void
}

export function Game({ gameId, players, gameFinished }: GameProps) {
    const socket = useContext(SocketContext)
    const [winner, setWinner] = useState<User | null>(null)
    const [gameStarted, setGameStarted] = useState(false)
    const [mapData, setMapData] = useState<MapData | null>(null)
    const [lobbyOpen, setLobbyOpen] = useState(false)

    useEffect(() => {
        const onGameStarted = (response: GameState) => {
            setMapData(response.map)
            setGameStarted(true)
        }
        socket.on(IncomingEvents.GAME_STARTED, onGameStarted)

        const onGameEnd = (winner: User, players: User[]) => {
            setWinner(winner)
        }
        socket.on(IncomingEvents.GAME_END, onGameEnd)

        const onSync = (response: GameState) => {
            setMapData(response.map)
        }
        socket.on(IncomingEvents.SYNC, onSync)

        return () => {
            socket.off(IncomingEvents.GAME_STARTED, onGameStarted)
            socket.off(IncomingEvents.GAME_END, onGameEnd)
            socket.off(IncomingEvents.SYNC, onSync)
        }
    }, [socket, players])

    const startGame = () => {
        if (!gameStarted) {
            setGameStarted(true)
            socket.emit(OutgoingEvents.START_GAME, gameId)
        }
    }

    return (
        <>
            <div className="game">
                <header className="header">
                    <div className="game-id">
                        <b>Game ID: </b>
                        {gameId}
                    </div>
                    <button
                        id="start-game"
                        className="btn"
                        disabled={gameStarted || players.length < 2}
                        onClick={startGame}
                    >
                        START GAME
                    </button>
                    <button
                        id="lobby"
                        className="btn"
                        onClick={() => setLobbyOpen(true)}
                    >
                        LOBBY ({players.length})
                    </button>
                </header>
                <div className='game-map'>
                    {gameStarted && mapData && <GameMap mapData={mapData} gameId={gameId} />}
                </div>
            </div>
            {lobbyOpen && (
                <LobbyModal
                    players={players}
                    close={() => setLobbyOpen(false)}
                />
            )}
            {winner && (
                <SummaryModal
                    winner={winner}
                    close={gameFinished}
                />
            )}
        </>
    )
}
