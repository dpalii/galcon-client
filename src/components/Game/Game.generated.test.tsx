import React from 'react';

import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import { Game } from '../../../src/components/Game/Game';

let container: any = null;

beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div');
  document.body.appendChild(container);
});

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container);
  container.remove();
  container = null;
});

describe('<Game>', () => {
  it('should render component', () => {
    act(() => {
      render(<Game
        gameId={'/* string */'}
        players={[{
          name: 'name1', isHost: false, id: '1', color: '#ffffff',
        },
        {
          name: 'name2', isHost: true, id: '2', color: '#000000',
        },
        {
          name: 'name3', isHost: false, id: '3', color: '#fff000',
        }]}
        gameFinished={() => {}}
      />, container);
    });
    container.querySelector('#lobby').dispatchEvent(new MouseEvent('click', { bubbles: true }));
    container.querySelector('#start-game').dispatchEvent(new MouseEvent('click', { bubbles: true }));
    expect(container).toMatchSnapshot();
  });
});
