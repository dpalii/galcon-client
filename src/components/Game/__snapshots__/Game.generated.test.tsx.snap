// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Game> should render component 1`] = `
<div>
  <div
    class="game"
  >
    <header
      class="header"
    >
      <div
        class="game-id"
      >
        <b>
          Game ID: 
        </b>
        /* string */
      </div>
      <button
        class="btn"
        disabled=""
        id="start-game"
      >
        START GAME
      </button>
      <button
        class="btn"
        id="lobby"
      >
        LOBBY (
        3
        )
      </button>
    </header>
    <div
      class="game-map"
    />
  </div>
  <div
    class="backdrop"
  >
    <div
      class="modal"
    >
      <h2
        class="heading"
      >
        PLAYERS:
      </h2>
      <div
        class="player-entry"
      >
        <span
          class="player-color"
          style="background-color: rgb(255, 255, 255);"
        />
        <b>
          name1
           
        </b>
      </div>
      <div
        class="player-entry"
      >
        <span
          class="player-color"
          style="background-color: rgb(0, 0, 0);"
        />
        <b>
          name2
           
          (host)
        </b>
      </div>
      <div
        class="player-entry"
      >
        <span
          class="player-color"
          style="background-color: rgb(255, 240, 0);"
        />
        <b>
          name3
           
        </b>
      </div>
      <button
        class="btn w-100"
      >
        CLOSE
      </button>
    </div>
  </div>
</div>
`;
